### DISCUSSION TITLE
How to add menu item to help menu ?

### DISCUSSION THREAD

**Discussion by 80avin** [FreeCAD/FreeCAD#18773] [Discussions that will turn in to issues but need refining]
I am trying to build a command palette addon which can be triggered from the "Ctrl+Shift+P" shortcut or from an item under the Help menu.
However, it is becoming challenging to find how to add the menu item to the help menu.

### Problem description
Looking at https://github.com/FreeCAD/FreeCAD/blob/7083362acb8529017b58ea03bf7416da0192b25b/src/Gui/Workbench.cpp#L760 , I would expect that the following should find the menu
```python
Gui.getMainWindow().findChildren(QMenu, '&Help')
```
However, this returns 2 items.

At this point, I can use array indexing or some definite way to find the right QMenu object.
I don't want to use hardcoded indices as it may break in future. In trying to find a definite way, I found that following gives the right `QMenu`
```python
max(Gui.getMainWindow()..menuBar().findChildren(QMenu, "&Help"), key=lambda x: len(x.actions()))
```

However, this also breaks when I install [FreeCAD-Ribbon](https://github.com/APEbbers/FreeCAD-Ribbon), as the findChildren itself now returns 4-5 items.

While this can be translated as an issue in FreeCAD-Ribbon, it is essentially something lacking ( implementation or documentation ) in FreeCAD which is leading to this situation.

I have spent time in searching through the documentation and even github code search, but not able to find any clear soluton.

**Comment by luzpaz** [comment]
Running through copilot I was able to successfully (although not easily) figure out how to do it via python

```python
import FreeCADGui
from PySide2 import QtWidgets, QtCore

class MyHelpCommand:
    def GetResources(self):
        return {
            'MenuText': 'My Help Item',
            'ToolTip': 'This is my custom help item'
        }

    def Activated(self):
        QtWidgets.QMessageBox.information(None, "My Help Item", "This is my custom help item.")

# Register the command
FreeCADGui.addCommand('My_Help_Command', MyHelpCommand())
print("Command registered successfully")

def add_help_item():
    # Get the main window
    mw = FreeCADGui.getMainWindow()
    if mw is None:
        print("Main window not found")
        return

    print("Main window found")

    # Find the existing Help menu
    menu_bar = mw.menuBar()
    help_menu = None
    for action in menu_bar.actions():
        if action.text() == "&Help":
            help_menu = action.menu()
            break

    if help_menu is None:
        print("Help menu not found")
        return

    print("Help menu found")

    # Create a new action for the custom command
    new_action = QtWidgets.QAction('My Help Item', mw)
    new_action.triggered.connect(lambda: FreeCADGui.runCommand('My_Help_Command'))

    # Add the custom command to the Help menu
    help_menu.addAction(new_action)
    help_menu.repaint()
    QtCore.QCoreApplication.processEvents()
    print("Action added to the bottom of the Help menu")

# Ensure the action is added after the event loop has started
QtCore.QTimer.singleShot(1000, add_help_item)
```

**Comment by 80avin** [comment]
@luzpaz  Thanks, that works finally.
It looks like the issue arose because `QObject::findChildren` recursively finds all the items, while your approach only searches for direct children.

However, I wonder if is there any event/signal that I can listen to instead of using QTimer. By reading other implementations, I see people usually connect to the `FreeCADGui.getMainWindow().workbenchActivated` signal which can fire multiple times.

**Comment by yorikvanhavre** [comment]
Including items in any of the existing menus is best done with a workbench manipulator. See https://github.com/FreeCAD/FreeCAD/pull/10933

But you need first to create a FreeCAD command, as @luzpaz explained above

**Comment by 80avin** [comment]
@yorikvanhavre Thanks.
WorkbenchManipulator works to modify the menu consistently but brings new issues.
1. Reloading the workbench during init gives an error.
   ```
   File "<string>", line 94, in reloadActive
   AttributeError: 'PartDesignWorkbench' object has no attribute '__Workbench__'
   ```
2. If I reload it later using a timer, the menus added by other addons disappear.
    Example: PieMenu [adds an Accessories menu](https://github.com/Grubuntu/PieMenu/blob/4c494b8459db191b541c510e71aec0be72f78340/InitGui.py#L1829-L1841) which gets removed when I run `reloadActive` later.

<details>

<summary>
Click to see the code

</summary>

```python
fc_commandpalette_initialized = False


def should_initialize(wb_name):
    global fc_commandpalette_initialized
    if fc_commandpalette_initialized:
        return False
    if wb_name == "NoneWorkbench":
        return False
    fc_commandpalette_initialized = True
    return True


def init(wb_name):
    global should_initialize
    if not should_initialize(wb_name):
        return

    class MyHelpCommand:
        def GetResources(self):
            return {
                "MenuText": "Command Palette",
                "ToolTip": "Search Commands, menu, workbenches, properties, files, etc.",
            }

        def Activated(self):
            from PySide2 import QtWidgets

            QtWidgets.QMessageBox.information(
                None, "My Help Item", "This is my custom help item."
            )

    class WorkbenchManipulator:
        def modifyMenuBar(self):
            print("adding menu bars")
            return [
                {"insert": "mod_Command_Palette", "menuItem": "Std_FreeCADDonation"}
            ]

    FreeCADGui.addCommand("mod_Command_Palette", MyHelpCommand())
    FreeCADGui.addWorkbenchManipulator(WorkbenchManipulator())
    wb = FreeCADGui.activeWorkbench()
    wb.reloadActive()
    print("added wb")


import FreeCADGui

FreeCADGui.getMainWindow().workbenchActivated.connect(init)
```
</details>

I will look into the first issue but I guess FreeCAD's addon API needs some improvements.
For example, 
* Instead of WorkbenchManipulator, directly having functions under the `FreeCADGui` module to manipulate menus would be a better addition.
* A onLoaded ( or equivalent ) signal which runs only once, the entire startup has been completed.

**Comment by yorikvanhavre** [comment]
That might be because you are executing stuff before the GUI is fully loaded. One easy way to workaround that is to exectute stuff in a `QtCore.QTime.singleShot(0,my_function)`
